### PR TITLE
ISSUE-209 CalDavMultitenancyContract

### PR DESCRIPTION
**Pull request by hungphan227**

### CODE REVIEW

**Comment by chibenwa** [comment]
Please `@Disable` tests that are not passing with reference to the ticket before I review.

**Comment by hungphan227** [comment]
> Please `@Disable` tests that are not passing with reference to the ticket before I review.

could I use story ticket https://github.com/linagora/twake-calendar-side-service/issues/673 ?

**Comment by chibenwa** [comment]
yes

File: src/test/java/com/linagora/dav/contracts/cal/CalDavMultitenancyContract.java [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+                .add("Accept", "application/json")
+                .add("Content-Type", "application/json"))
+            .post()
+            .uri("/calendars/" + john.id() + "/" + john.id() + ".json")
+            .send(body("""
+                {"share":{"set":[],"remove":[]}}
+                """)));
+
+        assertThat(status).isEqualTo(403);
+    }
```
Comment (chibenwa) [inline src/test/java/com/linagora/dav/contracts/cal/CalDavMultitenancyContract.java:373]: I think this test is invalid, it sets what looks like ACL onto an event URL it looks odd

Can you have a look @hungphan227 ?

**Review by chibenwa — COMMENTED** [review]
@vttranlina please have a review of it too

@hungphan227 I am missing
 - free busy
 - counter 
 - please also try to think about missing tests

File: src/test/java/com/linagora/dav/contracts/cal/CalDavMultitenancyContract.java (hunk 1 above)
Comment (hungphan227) [reply src/test/java/com/linagora/dav/contracts/cal/CalDavMultitenancyContract.java:373]: yeah sr. updated

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
@vttranlina please have a read too

File: src/test/java/com/linagora/dav/contracts/cal/CalDavMultitenancyContract.java [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+        dockerExtension().twakeCalendarProvisioningService().createDomainIfNotExists(SECOND_DOMAIN);
+        bob = dockerExtension().newTestUser();
+        john = dockerExtension().twakeCalendarProvisioningService()
+            .createUser(UUID.randomUUID().toString(), SECOND_DOMAIN).block();
+    }
+
+    @Disabled("https://github.com/linagora/twake-calendar-side-service/issues/673")
+    @Test
+    void propfindShouldReturn403ForCrossDomainUser() {
+        calDavClient.updateCalendarAcl(john, "{DAV:}write");
```
Comment (vttranlina) [inline src/test/java/com/linagora/dav/contracts/cal/CalDavMultitenancyContract.java:67]: Should we move it to `BeforeEach` ?

File: src/test/java/com/linagora/dav/contracts/cal/CalDavMultitenancyContract.java [hunk 3]
```diff
@@ -0,0 +1,582 @@
...
+            .recipient(bob.email())
+            .uid(eventUid)
+            .method("COUNTER")
+            .buildJson();
+
+        assertThatThrownBy(() -> calDavClient.sendITIPRequest(john, URI.create("/calendars/" + bob.id()), counterBody).block())
+            .isInstanceOf(RuntimeException.class)
+            .hasMessageContaining("403");
+    }
+}
```
Comment (vttranlina) [inline src/test/java/com/linagora/dav/contracts/cal/CalDavMultitenancyContract.java:582]: Should we add tests for `resource`?

I’m not entirely sure it’s necessary, but `resource` seems to follow different logic/queries compared to regular users. Could you confirm?

File: src/test/java/com/linagora/dav/contracts/card/CardJsonContract.java [hunk 4]
```diff
@@ -1480,6 +1480,63 @@ void move() {
         assertThat(response2).contains("abcdef.vcf");
     }
 
+    @Test
+    void copy() {
```
Comment (vttranlina) [inline src/test/java/com/linagora/dav/contracts/card/CardJsonContract.java:1484]: Is this change related to this PR?

File: src/test/java/com/linagora/dav/contracts/cal/CalDavMultitenancyContract.java (hunk 2 above)
Comment (hungphan227) [reply src/test/java/com/linagora/dav/contracts/cal/CalDavMultitenancyContract.java:67]: Some tests does not require this.

File: src/test/java/com/linagora/dav/contracts/card/CardJsonContract.java (hunk 4 above)
Comment (hungphan227) [reply src/test/java/com/linagora/dav/contracts/card/CardJsonContract.java:1484]: When writing tests for cross domain, I found that we have not tested some methods of sabre. I added them to prove that they exists and actually work in single domain scope.

File: src/test/java/com/linagora/dav/contracts/cal/CalDavMultitenancyContract.java (hunk 3 above)
Comment (hungphan227) [reply src/test/java/com/linagora/dav/contracts/cal/CalDavMultitenancyContract.java:582]: IMO no

Comment (chibenwa) [reply src/test/java/com/linagora/dav/contracts/cal/CalDavMultitenancyContract.java:582]: That is a good catch: I would like us to test isolation on resources as well: it is a edge case.

File: src/test/java/com/linagora/dav/contracts/card/CardJsonContract.java (hunk 4 above)
Comment (chibenwa) [reply src/test/java/com/linagora/dav/contracts/card/CardJsonContract.java:1484]: Nice initiative BTW

**Comment by chibenwa** [comment]
I will complete this test suite:
 - Resource
 - base calendar listing
 - principals
